### PR TITLE
Format transaction notes as clickable tags

### PR DESCRIPTION
**Pull request by joel-jeremy**
For https://github.com/actualbudget/actual/issues/531

### CODE REVIEW

**Comment by youngcw** [comment]
The colors probably could use some work, but other than that this looks great

**Comment by joel-jeremy** [comment]
Any suggestions on the colors? Would it be better to apply primary colors to them? But that would make the tags stand out on the table.

**Comment by youngcw** [comment]
I don't think I have good recommendations for specific colors.  What I was seeing is that the pill colors can be hard to see when the transaction is hovered and highlighted (the same color in midnight mode), and the pill hover color was usually really close to the normal color so it was like nothing happened when hovering.

**Comment by youngcw** [comment]
Those colors look good to me

**Comment by Teprifer** [comment]
If you're already on the All Accounts screen, clicking the tag doesn't apply the filtering.

**Comment by joel-jeremy** [comment]
> If you're already on the All Accounts screen, clicking the tag doesn't apply the filtering.
Should be fixed now :)

**Comment by Teprifer** [comment]
Thanks! :)

I also like the new behaviour where it only filters the current account screen when clicked on, I think it's more intuitive.

**Comment by MatissJanis** [comment]
Maybe it's just me, but I dislike that the tags are in bold. Nothing else in the table is in bold. 

Other than that: this is an amazing addition! Thanks for working on this!

Screenshot 2024-04-27 at 22 00 47

File: packages/desktop-client/src/components/transactions/TransactionList.jsx [hunk 1]
```diff
@@ -161,13 +164,19 @@ export function TransactionList({
...
   const onNavigateToSchedule = useCallback(scheduleId => {
-    pushModal('schedule-edit', { id: scheduleId });
+    dispatch(pushModal('schedule-edit', { id: scheduleId }));
+  });
+
+  const onNavigateToFilteredTagView = useCallback(noteTag => {
+    const filterConditions = [
+      { field: 'notes', op: 'contains', value: noteTag, type: 'string' },
+    ];
+    onApplyFilters(filterConditions);
```
Comment (MatissJanis) [inline packages/desktop-client/src/components/transactions/TransactionList.jsx:174]: **💭 thought:** ‏currently the click on a tag REPLACES the existing filters. Would it be better to instead `append` a filter?

I can see a use case where I filter by category. And then click on a tag to narrow it down even further. Currently that would remove the category filter.

WDYT?

File: packages/desktop-client/src/components/transactions/TransactionsTable.jsx [hunk 2]
```diff
@@ -2172,3 +2186,45 @@ export const TransactionTable = forwardRef((props, ref) => {
 });
 
 TransactionTable.displayName = 'TransactionTable';
+
+function notesTagFormatter(value, onNoteTagClick) {
+  const words = value.split(' ');
+  return (
+    <>
```
Comment (MatissJanis) [inline packages/desktop-client/src/components/transactions/TransactionsTable.jsx:2193]: **🥜 nitpick:** ‏unnecessary wrapping fragment.

**Review by MatissJanis — DISMISSED** [review]

File: packages/desktop-client/src/style/themes/dark.ts [hunk 3]
```diff
@@ -176,11 +176,11 @@ export const checkboxBorderSelected = colorPalette.purple300;
 export const checkboxShadowSelected = colorPalette.purple500;
 export const checkboxToggleBackground = colorPalette.gray700;
 
-export const pillBackground = colorPalette.navy800;
+export const pillBackground = colorPalette.navy600;
```
Comment (carkom) [inline packages/desktop-client/src/style/themes/dark.ts:179]: This change about 4 other pages as well. It also doesn't pass AA.

**Review by carkom — CHANGES_REQUESTED** [review]
First of all this is amazing and I love the functionality. Any chance we could add an autocomplete list to the tags?

There's a lot going on in this PR. 

I'm not sure why changing naming conventions for filters across the entire app is relevant for this PR, would be better as a separate PR.

Color changes are no good. They are used in 4 other pages and the proposed changes do not pass AA (eg they are harder to read). Might also be best in a different PR as the pill color doesn't really relate to the core functionality of this PR.

**Comment by youngcw** [comment]
If tags have overlap in their names then you get multiple tags in the filter.  Example makes tags for `#tag` and `#tag2` and then click on `#tag` and both show up in the filtered list

File: packages/desktop-client/src/style/themes/dark.ts (hunk 3 above)
Comment (joel-jeremy) [reply packages/desktop-client/src/style/themes/dark.ts]: Nice catch! I forgot to revert those back. I tried reusing pill colors initially but decided to use different theme variables.

I renamed some variables because it was a bit hard to follow when we interchange filters with conditions throughout the codebase. In my mind, they are 2 different things/concepts - filters = saved filters (which uses conditions), conditions = rule conditions (but also used by filters). Just thought it would be more consistent to rename some of these variables.

File: packages/desktop-client/e2e/transactions.test.js-snapshots/Transactions-filters-transactions-by-category-7-chromium-linux.png
Comment (youngcw) [inline packages/desktop-client/e2e/transactions.test.js-snapshots/Transactions-filters-transactions-by-category-7-chromium-linux.png:1]: Why are these updating?  The "X" on the filter pill is moving.  Was that on purpose?

Comment (joel-jeremy) [reply packages/desktop-client/e2e/transactions.test.js-snapshots/Transactions-filters-transactions-by-category-7-chromium-linux.png:1]: Yes, this is intentional. If you open the demo version and highlight the filter pill and the X button, they overlap.

File: packages/desktop-client/src/style/themes/dark.ts (hunk 3 above)
Comment (carkom) [reply packages/desktop-client/src/style/themes/dark.ts]: Yea, I like the changes. I agree with you sentiment - lots of intermixing of terms. Was just commenting that it's a bit outside the scope of the PR. No worries, though!

File: packages/desktop-client/src/components/transactions/TransactionList.jsx [hunk 4]
```diff
@@ -161,13 +164,19 @@ export function TransactionList({
   });
 
   const onNavigateToSchedule = useCallback(scheduleId => {
-    pushModal('schedule-edit', { id: scheduleId });
+    dispatch(pushModal('schedule-edit', { id: scheduleId }));
+  });
+
+  const onNavigateToFilteredTagView = useCallback(noteTag => {
+    const filterConditions = [
+      { field: 'notes', op: 'contains', value: noteTag, type: 'string' },
```
Comment (carkom) [inline packages/desktop-client/src/components/transactions/TransactionList.jsx:172]: Should this be `op: 'is'` instead to pull only unique tag names?

Comment (joel-jeremy) [reply packages/desktop-client/src/components/transactions/TransactionList.jsx:172]: Tried experimenting with using `is`. The problem with using it is that it won't be able to support multiple tags and non-tag texts (no #) in the notes.

Comment (carkom) [reply packages/desktop-client/src/components/transactions/TransactionList.jsx:172]: Can you give specific examples of these? Eg. `notes contains ...`

Comment (joel-jeremy) [reply packages/desktop-client/src/components/transactions/TransactionList.jsx:172]: You can try adding a note with a tag and regular note e.g. `A sample #tag with a regular note`. If you click on the tag it won't match the transaction.

Comment (carkom) [reply packages/desktop-client/src/components/transactions/TransactionList.jsx:172]: Okay, I see now. So the PR isn't actually creating tags or keeping a list of created tags. It's just formatting the notes to give a pill around anything starting with #. So this is primarily a visual change not a functional addition.

Would be good (in the future) to make tags an actual saved item (new db table) so we can manage the tags and enable autocomplete, etc.

**Comment by joel-jeremy** [comment]
> If tags have overlap in their names then you get multiple tags in the filter. Example: make tags for #tag and #tag2 and then click on #tag and both show up in the filtered list

This would be the limitation of using the existing notes as basis for tags. Until we decide to save tags in the database in their own fields, this would probably be the best we can do for now

**Comment by youngcw** [comment]
It works for me to put a space at the end of the "contains" search in the filter.  That works as long as there is a space after the tag in the note

**Comment by joel-jeremy** [comment]
> It works for me to put a space at the end of the "contains" search in the filter. That works as long as there is a space after the tag in the note. The case that doesn't work is a single tag in a note that works as a tag, but doesn't have a space at the end

Yeah, I was getting this too. It also won't work if the tag is in the end of the notes.

**Comment by youngcw** [comment]
It could be nice to add a special filter that looks for tags that has some extra regex in it than just a "contains" search.  Could check for either a space or a line end to match tags.  That would also be more clear for when making filters and users want to search for tags.

File: packages/desktop-client/e2e/transactions.test.js-snapshots/Transactions-filters-transactions-by-category-7-chromium-linux.png
Comment (youngcw) [reply packages/desktop-client/e2e/transactions.test.js-snapshots/Transactions-filters-transactions-by-category-7-chromium-linux.png:1]: Could we make the padding on each side of the filter pill match still?  With this change the padding on the "X" side is smaller than the left side.

**Comment by youngcw** [comment]
Are we thinking this will get set as experimental or do we want try to get a solid MVP before merging?  I think we are nearly there with an MVP

**Comment by youngcw** [comment]
This will need a call out in the docs somewhere

**Comment by psybers** [comment]
> It could be nice to add a special filter that looks for tags that has some extra regex in it than just a "contains" search. Could check for either a space or a line end to match tags. That would also be more clear for when making filters and users want to search for tags.

Couldn't you set the filter value to something like `%[^a-z0-9-]TAGHERE[^a-z0-9-]%`? I am making the assumption that a "tag" is just letters, digits, and hyphens.

https://github.com/actualbudget/actual/pull/2670/files#diff-b6fb705e038031b12b7b1d3065b3889cd7b36a0e7668a73d306ddc1a35a1a6a4R172

**Comment by youngcw** [comment]
Looks like regex's `#[\S]*\b` for finding the tag and `#tag\b` for filtering by the tag could work.  That would allow anything in the tag, even more `#` chars.

**Comment by psybers** [comment]
> Looks like regex's `#[\S]*\b` for finding the tag and `#tag\b` for filtering by the tag could work. That would allow anything in the tag, even more `#` chars.

Yes, I was aiming for the word boundary regex.  But isn't this just going to a SQL statement `WHERE foo LIKE %tag%`?  I don't believe we could use regex there?

**Comment by psybers** [comment]
To be honest, I am not sure you can even put multiple ranges into a `LIKE` pattern.  I know you can do a character range (or the negation) but have never tried multiple ranges.

**Comment by joel-jeremy** [comment]
> Yes, I was aiming for the word boundary regex. But isn't this just going to a SQL statement WHERE foo LIKE %tag%? I don't believe we could use regex there?

Yes, this will all be compiled down to a LIKE %TAG% query

**Comment by psybers** [comment]
Ok it appears `LIKE` does not even support character classes/ranges.  I think I have confused using MySQL's `RLIKE` operator in the past.

The `LIKE` operator only supports `%` and `_` wildcards.

You could get a bit hacky and work around it though.  Something like this might work:

```sql
SELECT * FROM table WHERE CONCAT(' ', notes, ' ') LIKE '% #tag %';
```

Then if the tag is at the start or end, it would still match.  This does assume that tags are surrounded by spaces (not newlines or tabs).

Another option is to install a regexp module for sqlite (https://www.npmjs.com/package/sqlite-regex) and then use that to match.

**Comment by youngcw** [comment]
Could something like this work? https://github.com/WiseLibs/better-sqlite3/issues/763

We would add a regex function that can be run inside a query.

**Comment by youngcw** [comment]
Nice!  The only thing off thing I'm noticing is that the `\b` is showing in the filter pill with the tag name.

**Comment by joel-jeremy** [comment]
I have implemented a `$regexp` operator in aql and used that in rule's `matches` op.

Hopefully this also paves the way to implement regex in rules as per: https://github.com/actualbudget/actual/issues/500

> Nice! The only off thing I'm noticing is that the \b is showing in the filter pill with the tag name.

The `matches` filter should display the raw regex that is being matched so I think it's fine having that there. We could consider adding a special Tags filter if we want in the future to display the just tags.

**Comment by psybers** [comment]
Still not quite working as you would expect:

image
image

**Comment by youngcw** [comment]
> The `matches` filter should display the raw regex that is being matched so I think it's fine having that there. We could consider adding a special Tags filter if we want in the future to display the just tags.

I think a dedicated tag filter would be good, but that could be a later addition.  Thanks for all your work on this!

**Comment by joel-jeremy** [comment]
> Still not quite working as you would expect:

This matched the transaction because the regex matched the #test tag in the notes

**Comment by psybers** [comment]
> > Still not quite working as you would expect:
> 
> This matched the transaction because the regex matched the #test tag in the notes

Yes, you are right. I've been running on 3-5 hours of sleep a night for a few weeks...

One thing it would match, that maybe we do not want(?) would be something like "#this#is#a#test".  Do we want a `/b` on the left?

**Comment by joel-jeremy** [comment]
@youngcw @psybers Putting \b on the start of the regex expression doesn't seem to work as expected. Any suggestions?

**Comment by youngcw** [comment]
@joel-jeremy What about something like `#[^\s#]*\b`  That will catch everything but # or whitespace

**Comment by joel-jeremy** [comment]
@youngcw That's works great when for searching all tags. But it doesn't seem to work when searching for a single tag.

**Comment by youngcw** [comment]
> That's works great when for searching all tags. But it doesn't seem to work when searching for a single tag.

Wouldn't the existing #tag/b still work if that other expression was used to mark the tags?

**Comment by psybers** [comment]
I think the regex needs to be something like `(^|\s)#Test\b`.  So it ends on a word boundary but starts at either the beginning of the note or a space.

**Comment by youngcw** [comment]
I guess maybe we need to nail down what we want to have count as a tag and what is allowed in a tag.  After that its pretty simple to make a regex to match those rules we want to enforce

**Comment by joel-jeremy** [comment]
> I think the regex needs to be something like (^|\s)#Test\b. So it ends on a word boundary but starts at either the beginning of the note or a space.

This seem to work great:
image

> I guess maybe we need to nail down what we want to have count as a tag and what is allowed in a tag. After that its pretty simple to make a regex to match those rules we want to enforce

I would say we should just allow any characters in tags

**Comment by psybers** [comment]
Seems to be a bug.  If you select two tags, it is set to `all` yet it isn't behaving that way.

bug

If you change it to `any` and back to `all` you see the transaction list is not the same.

**Comment by psybers** [comment]
Also, weird tags break things:

image

Clicking that you get nothing matched.

**Comment by psybers** [comment]
Probably it is because of the regex characters in it - you might need to escape the tag when building the regex?

**Comment by joel-jeremy** [comment]
@psybers I have pushed a commit which escapes the regex. Tested some special characters and it seem to work fine expect when using `)`. It 
 won't show any results if tag contains `)` for some reason, even if it's escaped.

**Comment by psybers** [comment]
So the question still remains, is this what we want or not?

image

With `\b` on the right side, it will match `#test#that` (because '#' is a word boundary).

**Comment by psybers** [comment]
If we want only "whole tags" and consider that to mean whitespace delineated, then the regex could be `(^|\s)#test($|\s)`.

**Comment by joel-jeremy** [comment]
> If we want only "whole tags" and consider that to mean whitespace delineated, then the regex could be (^|\s)#test($|\s).

Yeah, when clicking tags, this is how I would expect the filters to behave. `(^|\s)#test($|\s)` seems to work well.

**Comment by youngcw** [comment]
I still don't like that `#test#tag` would count as one long tag.  It probably would make more sense to have that parse as not a tag if we are going to assume a space at the end for the search.

**Comment by psybers** [comment]
So there are three options for `#test#tag`:

1. Treat it as two, separate tags.
2. Treat it as a single big tag.
3. Ignore it. It doesn't look like a tag, so don't turn it into a button.

Is there a commonly accepted answer used by many other software?  If so I would say mimic the common behavior.  If not, my vote is for 1.

The current solution is 2.  If we decide to switch to one of the other solutions, the regex will probably need updated too.

**Comment by youngcw** [comment]
> So there are three options for `#test#tag`:
> 
> 1. Treat it as two, separate tags.
> 2. Treat it as a single big tag.
> 3. Ignore it. It doesn't look like a tag, so don't turn it into a button.

I would vote for option 1.  I think that would be the most obvious to people and we can easily parse that with /b in a regex.  Finding the tags should be pretty straight forward to.  Look for a `#`, end the tag when there is another `#` or a space.  The searching regex would have to escape most symbols if they are allowed

**Comment by psybers** [comment]
For what its worth, 1 is how Facebook behaves.
image
This is two separate tags.

**Comment by joel-jeremy** [comment]
@psybers @youngcw Pushed a change to treat as separate tags

**Comment by psybers** [comment]
This behavior seems good to me.

**Comment by youngcw** [comment]
That seems to be working really well.  

Some things of note:
* `$` in a tag breaks the search.  I can't tell why since it does get escaped.  All other symbols I have seem to work fine
* A single `#` at the end of a tag disappears, not the end of the world, but a bit strange.  It doesn't show up as a tag, which is fine, but it doesn't show in the rendered note either.

**Comment by psybers** [comment]
Here is a weird one: `test#`

Turns it into `#test`.

**Comment by psybers** [comment]
And possibly related? `foo#bar` becomes `#foo #bar`

**Comment by joel-jeremy** [comment]
Updated the rendering logic of tags. The current regex doesn't seem work for tags that are in the middle of a string.

**Comment by psybers** [comment]
Might have to allow it to be bordered by another tag... maybe something like this?

`(^|\s|\w)#test($|\s|#)`

**Comment by joel-jeremy** [comment]
@psybers The problem string is when tag is preceded by a `#` e.g. `Onelongwordinnoteswith##tag`. This is not being matched by the current and the new regex you provided.

**Comment by psybers** [comment]
I mean the simple fix is to add `#` on the left side.  `(^|\s|\w|#)#test($|\s|#)`

But this regex is feeling more and more hacky and I worry it will match things we do not intend.  The best approach would be to know a more or less complete set of all cases we want to match (and cases to avoid), build the NFA by hand, and then turn it into a regex.

**Comment by psybers** [comment]
I have thought more about this and am wondering if we need the left side at all.  The regex will find things like `#test` and then enforce the right side sort of terminates the tag, either by whitespace, the end of the note, or another `#` character to start another tag.

But the left shouldn't really matter.  What comes prior to the `#test` does not matter.  It could be punctuation, letters, spaces, anything really.

**Comment by jfdoming** [comment]
> I have thought more about this and am wondering if we need the left side at all. The regex will find things like `#test` and then enforce the right side sort of terminates the tag, either by whitespace, the end of the note, or another `#` character to start another tag.
> 
> But the left shouldn't really matter. What comes prior to the `#test` does not matter. It could be punctuation, letters, spaces, anything really.

Not sure if _anything_ works, for example you could have a link with a hash like https://google.com#test

**Comment by psybers** [comment]
> > I have thought more about this and am wondering if we need the left side at all. The regex will find things like `#test` and then enforce the right side sort of terminates the tag, either by whitespace, the end of the note, or another `#` character to start another tag.
> > But the left shouldn't really matter. What comes prior to the `#test` does not matter. It could be punctuation, letters, spaces, anything really.
> 
> Not sure if _anything_ works, for example you could have a link with a hash like https://google.com#test

@jfdoming That is a fun edge case!  We could definitely filter out URLs if we wanted.

**Comment by psybers** [comment]
@jfdoming For what it's worth, your test case isn't turning into a tag currently:

image

**Comment by joel-jeremy** [comment]
Nice! More testing would be very much appreciated! Any other things left to address on this PR?

**Comment by youngcw** [comment]
> @jfdoming For what it's worth, your test case isn't turning into a tag currently:
> 
> image

That url turns into a tag for me.  just the last part though.  I don't know if there would be a good way around that.  Maybe it could work to only allow no space before a tag if the `#` was preceded by another tag.  I don' think that is a deal breaker right now.

**Comment by psybers** [comment]
> That url turns into a tag for me. But just the last, tag part. I don't know if there would be a good way around that. Maybe it could work to only allow no space before a tag if the `#` was preceded by another tag. I don' think that is a deal breaker right now.

https://stackoverflow.com/questions/63845715/regex-dont-match-if-match-is-inside-a-url

**Review by psybers — CHANGES_REQUESTED** [review]

**Comment by youngcw** [comment]
Im getting this error when clicking on a tag
```
Uncaught (in promise) aW: Field “field” does not exist in table “transactions”

Expression stack:
  $field
  {"field":{"$eq":"notes"}}
  filter([
    {
      "$and": [
        {
          "field": "notes",
          "op": "matches",
          "value": "(^|\\s|\\w|#)#tag($|\\s|#)",
          "type": "string"
        }
      ]
    },
    {
      "account": "b4592337-3a02-4d7e-b93c-90c734b6b91f"
    }
  ])
```

**Comment by joel-jeremy** [comment]
> Im getting this error when clicking on a tag

This should be fixed now

File: packages/desktop-client/e2e/transactions.test.js-snapshots/Transactions-creates-a-split-test-transaction-1-chromium-linux.png
Comment (youngcw) [inline packages/desktop-client/e2e/transactions.test.js-snapshots/Transactions-creates-a-split-test-transaction-1-chromium-linux.png:1]: Why did that filter pill pop up?

File: packages/desktop-client/e2e/transactions.test.js-snapshots/Transactions-filters-transactions-by-date-7-chromium-linux.png
Comment (youngcw) [inline packages/desktop-client/e2e/transactions.test.js-snapshots/Transactions-filters-transactions-by-date-7-chromium-linux.png:1]: Something is broken here where the filter pill is not rendering.

**Comment by joel-jeremy** [comment]
@youngcw The filtered balance should be good now

**Comment by psybers** [comment]
@joel-jeremy Did you see that other PR for this feature?  They had it so there was a "tag:" filter.  Any reason we can't do that, instead of the weird regex stuff in this PR?

Other things I liked in that PR: the tags rendered as you typed them out.  Tags also rendered in the filter box.

**Comment by psybers** [comment]
I think a lot of the rendering things can be fixed later.  But I wonder if the regex thing should even go in if there is a way to avoid it at all.

**Comment by youngcw** [comment]
I think that plenty of people will like having the matches filter available, but I would also like there to be a dedicated tag filter.  Currently if you are trying to filter for a tag you basically need to click on one to create the filter for you instead of being able to just start with making a filter yourself.

**Comment by carkom** [comment]
I've always thought it would be quite nice to have a db table of saved tags which would enable an auto complete list for both adding tags as well as searching/filtering them. Could all be managed within the auto complete menu similar to payees that has a "create new payee" option. No need for another payees type page (although could do if we wanted). 

I can rarely remember if I used a dash or an underscore for my tags. Looking it up is a pain.

**Comment by joel-jeremy** [comment]
Maybe limiting this PR to the matches filter is the way to go here? We could think about introducing tags filter on a separate PR.

**Review by youngcw — APPROVED** [review]
Im ok with this for now, but I think we need either some additions or a feature flag added before the release.  The things that I think need added soon are:
* A dedicated tag filter
* Unify what the filter accepts as a tag with what visually counts as a tag (something like "https://google.com/page#header" picks up the last part as a tag, but you cant find that part in a search)

**Comment by MatissJanis** [comment]
Are we good to merge this? :) 

cc @carkom your review shows up as "changes requested". If that's no longer applicable - would you mind removing the review (or better yet - doing a fresh review)? Just so we're more explicit about the state of this PR. Very much appreciated! 🙇

**Comment by carkom** [comment]
Just a heads up. This has created a regression in reports. Variables like  `filters` and `conditions` are universally used variables. If we are going to change variable names in the accounts file, we need to make sure to check/change all files in the app that use that variable. I've put a fix into PR #2871 . Cheers!
